druid/src/widget/wgpu_view.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright 2018 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A view hosting a wgpu instance for 3D rendering.
+//!
+//! This widget has several limitations compared to other standard widget due
+//! to the way it interfaces with wgpu via its own native window, instead of
+//! reusing the same top-level native window like other widgets do. In particular
+//! this means it can only paint within its bounds and cannot be painted over by
+//! any other widget, and therefore does not support being embedded into other
+//! widgets like scroll views which partially obscure a widget's client area.
+
+#![cfg(feature = "wgpu_view")]
+
+use crate::widget::prelude::*;
+use crate::{Data, NativeWindowHandle};
+
+use log::{debug, info};
+
+/// A trait for rendering custom content into a `WgpuView` by accessing its internal wgpu pipeline.
+pub trait WgpuRenderer {
+    /// Callback invoked to initilize the painter once the wgpu device and swap chain are up.
+    fn init(
```

Comment:
I *think* this would be better if we passed a function type that can construct the WGpuRenderer itself. 
Otherwise every implementer of the trait needs to handle the 'unitialised' state, by being an enum or containing an option.

Reply:
That makes sense a priori. I didn't touch that code for a few weeks so I'm not sure I fully grasp that but I can give it a try for sure.